### DISCUSSION TITLE
Proposal: conditional compilation symbols/pre-processor: allow additional operators

### DISCUSSION THREAD

**Discussion by jjxtra** [dotnet/csharplang#3254] [Language Ideas]
This would allow >=, <=, < and >. Preprocessor value would be converted to long or double if needed. Would be helpful comparing versions or other values that are not a simple true/false.

Use case: Unity engine defines preprocessor such as 2019_3_OR_NEWER, and does for each version of the editor. I’m sure there are many other use cases as well.

Example of new feature:

In project conditional compilation symbols:

`UNITY_VERSION_MAJOR=2019; UNITY_VERSION_MINOR=3; UNITY_VERSION_PATCH=36`

Code:
```csharp
#if UNITY_VERSION_MAJOR >= 2019 && UNITY_VERSION_MINOR >= 3
...
#endif
```

If this has been suggested before feel free to close otherwise I can tidy it up to match the feature suggestion template.

**Comment by theunrepentantgeek** [comment]
In the early days of .NET, I remember reading blog posts about why the C# preprocessor was quite deliberately simplistic - but I couldn't find them with a quick search.

> Preprocessor value would be converted to int or float if needed

This wouldn't work. 

Version **2019.11** is more recent than version **2019.8**, but a floating point comparison would get them the wrong way around.

**Comment by jjxtra** [comment]
Version could be zero padded on the minor and patch, like 201900030036. Would need to use a long but that’s ok

**Comment by HaloFour** [comment]
> Version could be zero padded on the minor and patch, like 201900030036. Would need to use a long but that’s ok

Zero padded out how many zeros?  `Minor` can be a signed 32-bit value, so you'd need 9 zeros.  Another 9 for the patch.  You've already exceeded the precision of a `double` at that point.  I don't think you could treat a version as a single preprocessor value even if that was allowed.

The limitations on preprocessor directives was intentional here.  It may be interesting to see if being limited to symbols has been annoying for the CoreFX team.

**Comment by jjxtra** [comment]
> > Version could be zero padded on the minor and patch, like [201900030036](tel:201900030036). Would need to use a long but that’s ok
> 
> Zero padded out how many zeros? `Minor` can be a signed 32-bit value, so you'd need 9 zeros. Another 9 for the patch. You've already exceeded the precision of a `double` at that point. I don't think you could treat a version as a single preprocessor value even if that was allowed.
> 
> The limitations on preprocessor directives was intentional here. It may be interesting to see if being limited to symbols has been annoying for the CoreFX team.

That’s why I mentioned using a long. I updated the proposal example with something that is cleaner.

**Comment by CyrusNajmabadi** [comment]
If this was a major ecosystem ask, i think it could be done.  However, touching this corner of the language for something not common is pretty unlikely to happen.

**Comment by jjxtra** [comment]
> 
> 
> If this was a major ecosystem ask, i think it could be done. However, touching this corner of the language for something not common is pretty unlikely to happen.

Perhaps it is a niche feature, more data would need to be gathered. I am hopeful it would be a fairly simple tweak to the C# pre-processor logic, but I could be wrong.
